File: app/Sources/FluesterungKit/GitHubDeviceAuth.swift
import Foundation

/// GitHub OAuth device flow — the only officially secretless flow: just a
/// client ID, no callback, no server. The token is used once for the profile
/// fetch and then discarded; nothing account-like is ever stored.
public struct GitHubDeviceCodeGrant: Sendable, Equatable {
    public let deviceCode: String
    public let userCode: String
    public let verificationURI: URL
    public let expiresAt: Date
    public let interval: TimeInterval

    public init(
        deviceCode: String,
        userCode: String,
        verificationURI: URL,
        expiresAt: Date,
        interval: TimeInterval
    ) {
        self.deviceCode = deviceCode
        self.userCode = userCode
        self.verificationURI = verificationURI
        self.expiresAt = expiresAt
        self.interval = interval
    }
}

public struct GitHubUser: Sendable, Equatable {
    public let login: String
    public let name: String?
    public let avatarURL: URL?

    public init(login: String, name: String?, avatarURL: URL?) {
        self.login = login
        self.name = name
        self.avatarURL = avatarURL
    }
}

public enum GitHubAuthError: Error, Equatable {
    /// "Enable Device Flow" is not ticked in the OAuth app settings.
    case deviceFlowDisabled
    /// The 15-minute device-code window elapsed.
    case expired
    /// The user cancelled on github.com.
    case accessDenied
    case http(Int)
    case malformedResponse
}

public struct GitHubDeviceAuth: Sendable {
    public typealias Transport = @Sendable (URLRequest) async throws -> (Data, URLResponse)
    public typealias Sleep = @Sendable (TimeInterval) async throws -> Void

    private let clientID: String
    private let transport: Transport
    private let sleep: Sleep
    private let now: @Sendable () -> Date

    /// Ephemeral on purpose: URLSession.shared's URLCache may write the
    /// bearer-authorized /user response to disk — this keeps every GitHub
    /// exchange RAM-only, preserving "the token is never persisted".
    private static let session = URLSession(configuration: .ephemeral)

    public init(
        clientID: String,
        transport: Transport? = nil,
        sleep: @escaping Sleep = { try await Task.sleep(for: .seconds($0)) },
        now: @escaping @Sendable () -> Date = Date.init
    ) {
        self.clientID = clientID
        self.transport = transport ?? { try await Self.session.data(for: $0) }
        self.sleep = sleep
        self.now = now
    }

    public func requestDeviceCode() async throws -> GitHubDeviceCodeGrant {
        let response = try await postForm(
            to: URL(string: "https://github.com/login/device/code")!,
            fields: ["client_id": clientID]
        )
        try Self.throwOnError(response)
        guard
            let deviceCode = response["device_code"] as? String,
            let userCode = response["user_code"] as? String,
            let verificationURI = (response["verification_uri"] as? String).flatMap(URL.init(string:)),
            let expiresIn = response["expires_in"] as? TimeInterval,
            let interval = response["interval"] as? TimeInterval
        else {
            throw GitHubAuthError.malformedResponse
        }
        return GitHubDeviceCodeGrant(
            deviceCode: deviceCode,
            userCode: userCode,
            verificationURI: verificationURI,
            expiresAt: now().addingTimeInterval(expiresIn),
            interval: interval
        )
    }

    /// Polls until the user authorizes on github.com. Cancellation propagates
    /// out of the injected sleep, so callers can abort cleanly.
    public func pollForAccessToken(_ grant: GitHubDeviceCodeGrant) async throws -> String {
        var interval = max(grant.interval, 1)
        while true {
            try await sleep(interval)
            guard now() < grant.expiresAt else { throw GitHubAuthError.expired }

            let response = try await postForm(
                to: URL(string: "https://github.com/login/oauth/access_token")!,
                fields: [
                    "client_id": clientID,
                    "device_code": grant.deviceCode,
                    "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
                ]
            )
            if let token = response["access_token"] as? String {
                return token
            }
            switch response["error"] as? String {
            case "authorization_pending":
                continue
            case "slow_down":
                interval = (response["interval"] as? TimeInterval) ?? (interval + 5)
            case "expired_token":
                throw GitHubAuthError.expired
            case "access_denied":
                throw GitHubAuthError.accessDenied
            case "device_flow_disabled":
                throw GitHubAuthError.deviceFlowDisabled
            default:
                throw GitHubAuthError.malformedResponse
            }
        }
    }

    public func fetchUser(token: String) async throws -> GitHubUser {
        var request = URLRequest(url: URL(string: "https://api.github.com/user")!)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("2022-11-28", forHTTPHeaderField: "X-GitHub-Api-Version")
        request.setValue("fluesterung", forHTTPHeaderField: "User-Agent")

        let json = try await send(request)
        guard let login = json["login"] as? String else {
            throw GitHubAuthError.malformedResponse
        }
        return GitHubUser(
            login: login,
            name: json["name"] as? String,
            avatarURL: (json["avatar_url"] as? String).flatMap(URL.init(string:))
        )
    }

    /// Raw avatar bytes, requested at the given pixel size. The size goes in
    /// as an extra query item — avatar URLs already carry `?v=4`.
    public func fetchAvatar(from url: URL, pixelSize: Int) async throws -> Data {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw GitHubAuthError.malformedResponse
        }
        components.queryItems = (components.queryItems ?? [])
            + [URLQueryItem(name: "s", value: String(pixelSize))]
        guard let sizedURL = components.url else { throw GitHubAuthError.malformedResponse }

        var request = URLRequest(url: sizedURL)
        request.setValue("fluesterung", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await transport(request)
        if let http = response as? HTTPURLResponse, http.statusCode != 200 {
            throw GitHubAuthError.http(http.statusCode)
        }
        return data
    }

    // MARK: - Internals

    private func postForm(to url: URL, fields: [String: String]) async throws -> [String: Any] {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("fluesterung", forHTTPHeaderField: "User-Agent")
        request.httpBody = Self.formEncode(fields).data(using: .utf8)
        return try await send(request)
    }

    private func send(_ request: URLRequest) async throws -> [String: Any] {
        let (data, response) = try await transport(request)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw GitHubAuthError.http(http.statusCode)
        }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw GitHubAuthError.malformedResponse
        }
        return json
    }

    /// GitHub answers some failures as HTTP 200 with an `error` field.
    private static func throwOnError(_ response: [String: Any]) throws {
        switch response["error"] as? String {
        case nil:
            return
        case "device_flow_disabled":
            throw GitHubAuthError.deviceFlowDisabled
        default:
            throw GitHubAuthError.malformedResponse
        }
    }

    private static func formEncode(_ fields: [String: String]) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return fields
            .sorted { $0.key < $1.key }
            .map { key, value in
                let encoded = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
                return "\(key)=\(encoded)"
            }
            .joined(separator: "&")
    }
}

File: apps/windows/src/core/github-device-auth.ts
import { getGitHubClientID } from '../main/github-config';

export interface GitHubDeviceCodeGrant {
	deviceCode: string;
	userCode: string;
	verificationURI: string;
	expiresAt: Date;
	intervalMs: number;
}

export interface GitHubUser {
	login: string;
	name?: string;
	avatarUrl?: string;
}

export type GitHubAuthError = 'device_flow_disabled' | 'expired' | 'access_denied' | 'http' | 'malformed';

export class GitHubAuthErrorResponse extends Error {
	constructor(
		public readonly code: GitHubAuthError,
		public readonly status?: number,
		message?: string,
	) {
		super(message ?? code);
		this.name = 'GitHubAuthErrorResponse';
	}
}

type FetchLike = typeof fetch;
type SleepLike = (ms: number) => Promise<void>;

interface GitHubDeviceAuthOptions {
	fetch?: FetchLike;
	sleep?: SleepLike;
	now?: () => Date;
}

const GITHUB_DEVICE_CODE_URL = 'https://github.com/login/device/code';
const GITHUB_ACCESS_TOKEN_URL = 'https://github.com/login/oauth/access_token';
const GITHUB_USER_URL = 'https://api.github.com/user';

export class GitHubDeviceAuth {
	private readonly clientID: string;
	private readonly fetchImpl: FetchLike;
	private readonly sleepImpl: SleepLike;
	private readonly now: () => Date;

	constructor(options: GitHubDeviceAuthOptions = {}) {
		this.clientID = getGitHubClientID();
		this.fetchImpl = options.fetch ?? fetch;
		this.sleepImpl = options.sleep ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
		this.now = options.now ?? (() => new Date());
	}

	async requestDeviceCode(): Promise<GitHubDeviceCodeGrant> {
		const response = await this.postForm(GITHUB_DEVICE_CODE_URL, [
			['client_id', this.clientID],
		]);
		this.throwOnTopLevelError(response, ['device_flow_disabled']);

		const deviceCode = typeof response.device_code === 'string' ? response.device_code : undefined;
		const userCode = typeof response.user_code === 'string' ? response.user_code : undefined;
		const verificationURI =
			typeof response.verification_uri === 'string' ? response.verification_uri : undefined;
		const expiresIn = this.readNumber(response.expires_in);
		const intervalSeconds = this.readNumber(response.interval);

		if (!deviceCode || !userCode || !verificationURI || expiresIn === undefined || intervalSeconds === undefined) {
			throw new GitHubAuthErrorResponse('malformed');
		}

		return {
			deviceCode,
			userCode,
			verificationURI,
			expiresAt: new Date(this.now().getTime() + expiresIn * 1000),
			intervalMs: intervalSeconds * 1000,
		};
	}

	async pollForAccessToken(grant: GitHubDeviceCodeGrant): Promise<string> {
		let intervalMs = Math.max(grant.intervalMs, 1000);

		while (true) {
			this.throwIfExpired(grant.expiresAt);
			await this.sleepImpl(intervalMs);
			this.throwIfExpired(grant.expiresAt);

			const response = await this.postForm(GITHUB_ACCESS_TOKEN_URL, [
				['client_id', this.clientID],
				['device_code', grant.deviceCode],
				['grant_type', 'urn:ietf:params:oauth:grant-type:device_code'],
			]);

			if (typeof response.access_token === 'string') {
				return response.access_token;
			}

			switch (response.error) {
				case 'authorization_pending':
					continue;
				case 'slow_down': {
					const serverIntervalMs = this.readNumber(response.interval);
					intervalMs = Math.max(
						intervalMs + 5000,
						serverIntervalMs === undefined ? 0 : serverIntervalMs * 1000,
					);
					continue;
				}
				case 'expired_token':
					throw new GitHubAuthErrorResponse('expired');
				case 'access_denied':
					throw new GitHubAuthErrorResponse('access_denied');
				case 'device_flow_disabled':
					throw new GitHubAuthErrorResponse('device_flow_disabled');
				default:
					throw new GitHubAuthErrorResponse('malformed');
			}
		}
	}

	async fetchUser(token: string): Promise<GitHubUser> {
		const response = await this.sendJson(GITHUB_USER_URL, {
			method: 'GET',
			headers: {
				Authorization: `Bearer ${token}`,
				Accept: 'application/vnd.github+json',
				'X-GitHub-Api-Version': '2022-11-28',
			},
		});

		const login = typeof response.login === 'string' ? response.login : undefined;
		if (!login) {
			throw new GitHubAuthErrorResponse('malformed');
		}

		return {
			login,
			name: typeof response.name === 'string' ? response.name : undefined,
			avatarUrl: typeof response.avatar_url === 'string' ? response.avatar_url : undefined,
		};
	}

	async fetchAvatar(url: string, pixelSize: number): Promise<Uint8Array> {
		const sizedUrl = new URL(url);
		sizedUrl.searchParams.append('s', String(pixelSize));

		const response = await this.fetchImpl(sizedUrl.toString(), {
			cache: 'no-store',
			headers: {
				'User-Agent': 'munkel',
			},
		});

		if (!response.ok) {
			throw new GitHubAuthErrorResponse('http', response.status);
		}

		const bytes = await response.arrayBuffer();
		return new Uint8Array(bytes);
	}

	private async postForm(url: string, entries: Array<[string, string]>): Promise<Record<string, unknown>> {
		const body = new URLSearchParams(entries).toString();
		return this.sendJson(url, {
			method: 'POST',
			headers: {
				Accept: 'application/json',
				'Content-Type': 'application/x-www-form-urlencoded',
			},
			body,
		});
	}

	private async sendJson(url: string, init: RequestInit): Promise<Record<string, unknown>> {
		const headers = new Headers(init.headers);
		headers.set('User-Agent', 'munkel');

		const response = await this.fetchImpl(url, {
			...init,
			cache: 'no-store',
			headers,
		});

		if (!response.ok) {
			throw new GitHubAuthErrorResponse('http', response.status);
		}

		let json: unknown;
		try {
			json = await response.json();
		} catch {
			throw new GitHubAuthErrorResponse('malformed');
		}

		if (!json || typeof json !== 'object' || Array.isArray(json)) {
			throw new GitHubAuthErrorResponse('malformed');
		}

		return json as Record<string, unknown>;
	}

	private throwOnTopLevelError(
		response: Record<string, unknown>,
		allowedErrors: GitHubAuthError[],
	): void {
		const error = typeof response.error === 'string' ? response.error : undefined;
		if (!error) {
			return;
		}
		if (allowedErrors.includes(error as GitHubAuthError)) {
			throw new GitHubAuthErrorResponse(error as GitHubAuthError);
		}
		throw new GitHubAuthErrorResponse('malformed');
	}

	private throwIfExpired(expiresAt: Date): void {
		if (this.now().getTime() >= expiresAt.getTime()) {
			throw new GitHubAuthErrorResponse('expired');
		}
	}

	private readNumber(value: unknown): number | undefined {
		return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
	}
}
